nix: downgrade from 2.14.1 to 2.13.4

jakubgs opened the pull request:
This Downgrades Nix interpreter from `2.14.1` to `2.13.4` to fix these errors:
* `error: creating directory '/nix/var/nix/gcroots/per-user/jakubgs': Permission denied`
* `error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)`

Resolves:
* https://github.com/status-im/status-mobile/issues/16344

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b4073414 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-16347/1/) | 2023-06-21 18:26:23 | ~6 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/p8bEpd) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2Fp8bEpd)|

jakubgs committed:
nix: downgrade from 2.14.1 to 2.13.4

Should resolve errors like this:
```
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)
```

Signed-off-by: Jakub Sokołowski <jakub@status.im>

siddarthkay commented on nix/scripts/setup.sh:
```diff
@@ -38,6 +40,33 @@ nix_install() {
+    else
+        echo "trusted-users = $USER" | sudo tee -a /etc/nix/nix.conf >/dev/null
+    fi
+    sudo systemctl restart nix-daemon
```
`systemctl` doesn't work for macOS to restart nix-daemon

```nix
if [[ "$(uname -s)" == "Darwin" ]]; then
        echo "Restarting Nix daemon Launchd service" >&2
        sudo launchctl unload /Library/LaunchDaemons/org.nixos.nix-daemon.plist
        sudo launchctl load   /Library/LaunchDaemons/org.nixos.nix-daemon.plist
elif [[ "$(uname -s)" == "Linux" ]] && [[ "$(nix_install_type)" == "multi" ]]; then
    echo "Restarting Nix daemon Systemd service" >&2
    systemctl daemon-reload
    systemctl restart nix-daemon
fi
```

siddarthkay replied:
After the `systemctl` fix.
`make run-clojure` fails with 

```log
ok      github.com/git-lfs/git-lfs/v3/lfs       4.831s
--- FAIL: TestDoWithAuthApprove (0.00s)
panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted [recovered]
        panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted

goroutine 8 [running]:
testing.tRunner.func1.2({0x13ea8c0, 0xc000192f60})
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/testing/testing.go:1399 +0x39f
panic({0x13ea8c0, 0xc000192f60})
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/runtime/panic.go:884 +0x212
net/http/httptest.newLocalListener()
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/net/http/httptest/server.go:71 +0xe9
net/http/httptest.NewUnstartedServer(...)
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/net/http/httptest/server.go:119
net/http/httptest.NewServer({0x152cea0?, 0xc000012930})
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/net/http/httptest/server.go:106 +0x2d
github.com/git-lfs/git-lfs/v3/lfsapi.TestDoWithAuthApprove(0xc0000e7a00)
        /private/tmp/nix-build-git-lfs-3.3.0.drv-0/source/lfsapi/auth_test.go:62 +0xc5
testing.tRunner(0xc0000e7a00, 0x14b4ce8)
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
        /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/testing/testing.go:1493 +0x35f
FAIL    github.com/git-lfs/git-lfs/v3/lfsapi    0.671s
FAIL
error: builder for '/nix/store/6pr30wwq3nz6vyvpiqc8if652ps5b44l-git-lfs-3.3.0.drv' failed with exit code 1;
       last 10 log lines:
       > net/http/httptest.NewServer({0x152cea0?, 0xc000012930})
       >         /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/net/http/httptest/server.go:106 +0x2d
       > github.com/git-lfs/git-lfs/v3/lfsapi.TestDoWithAuthApprove(0xc0000e7a00)
       >     /private/tmp/nix-build-git-lfs-3.3.0.drv-0/source/lfsapi/auth_test.go:62 +0xc5
       > testing.tRunner(0xc0000e7a00, 0x14b4ce8)
       >       /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/testing/testing.go:1446 +0x10b
       > created by testing.(*T).Run
       >         /nix/store/p2q9yykn507lh6hkvm5zl4nz0jxrlvd0-go-1.19.9/share/go/src/testing/testing.go:1493 +0x35f
       > FAIL github.com/git-lfs/git-lfs/v3/lfsapi    0.671s
       > FAIL
       For full logs, run 'nix-store -l /nix/store/6pr30wwq3nz6vyvpiqc8if652ps5b44l-git-lfs-3.3.0.drv'.
error: 1 dependencies of derivation '/nix/store/9h2b6nqyyqhw6182zif98lrvjjgvx6zq-nix-prefetch-git.drv' failed to build
error: 1 dependencies of derivation '/nix/store/h1l12lz8dgb7vjnjqvg67i9fd1r7jz85-yarn2nix-1.0.0.drv' failed to build
error:
       … while calling the 'derivationStrict' builtin

         at //builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'status-mobile-shell'
         whose name attribute is located at /nix/store/vcj90ksw5vk6ia960b73vpncxa9pw826-source/pkgs/stdenv/generic/make-derivation.nix:303:7

       … while evaluating attribute 'shellHook' of derivation 'status-mobile-shell'

         at /Users/siddarthkumar/code/status-mobile/nix/lib/mergeSh.nix:19:5:

           18|     # shellHook is a string, not a list
           19|     shellHook = (super.shellHook or "") + concatStrings (catAttrs "shellHook" shells);
             |     ^
           20|   })

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: 1 dependencies of derivation '/nix/store/psmqz8kw0kpx9wk09sn5fzc955jvd0fh-yarn.nix.drv' failed to build
make: *** [run-clojure] Error 1
```

jakubgs replied:
Yeah.

jakubgs replied:
You're mixing two issues in one comment.

siddarthkay commented:
purging and re-installing nix works fine on linux.
2 times I got this error below but it was fixed on a third try.
Most likely a network issue.

```log
patching sources
copying path '/nix/store/34lwx6v3g6x8g00c60fklq8lxvzj4nfs-openjdk-8u362-ga' from 'https://nix-cache.status.im'...
100 58.2M    0 58.2M    0     0   622k      0 --:--:--  0:01:35 --:--:--  441k
unpacking source archive /tmp/nix-build-status-go-fee033f-source.drv-0/fee033fadbaf7654ffde52c224e21603afbd122e.tar.gz
error: unable to download 'https://nix-cache.status.im/nar/ijpabfww24qfcb6fbjkbm184p5pfhm45.nar': HTTP error 200 (curl error: Stream error in the HTTP/2 framing layer)
error: some substitutes for the outputs of derivation '/nix/store/7y0aphnygd0x04y32x3mqsg1aa7hqpzq-androidsdk.drv' failed (usually happens due to networking issues); try '--fallback' to build derivation from source
error: 1 dependencies of derivation '/nix/store/3006mswjsygan8j37cyjqs9q4iim4p84-androidsdk-mod.drv' failed to build
error: build of '/nix/store/3006mswjsygan8j37cyjqs9q4iim4p84-androidsdk-mod.drv', '/nix/store/y5y47bckaqry07zhm06y49xnzvz6kmfb-status-go-0.159.2-fee033f-android.drv' failed
make: *** [Makefile:273: run-android] Error 1
siddarth@siddarth-ROG-Strix-GA35DX-G35DX:~/code/status-mobile$ make run-android
npx react-native run-android --appIdSuffix debug
Configuring Nix shell for target 'android'...
these 2 derivations will be built:
  /nix/store/3006mswjsygan8j37cyjqs9q4iim4p84-androidsdk-mod.drv
  /nix/store/y5y47bckaqry07zhm06y49xnzvz6kmfb-status-go-0.159.2-fee033f-android.drv
these 5 paths will be fetched (0.00 MiB download, 4101.06 MiB unpacked):
  /nix/store/n0imzzxf5p1r45j3bx9k5ibspdw2bmqw-android-sdk-build-tools-31.0.0
  /nix/store/ijpabfww24qfcb6fbjkbm184p5pfhm45-android-sdk-ndk-bundle-22.1.7171670
  /nix/store/mlqss2l2vqqpkr1advdgk1xai8iinm75-androidsdk
  /nix/store/rv8sqrqfs97c8rq18kx4isw63z7vxnc0-androidsdk
  /nix/store/rm4qwaznai98hg29d9b8w6fk3r5pqyln-gomobile-unstable-2022-05-18
copying path '/nix/store/ijpabfww24qfcb6fbjkbm184p5pfhm45-android-sdk-ndk-bundle-22.1.7171670' from 'https://nix-cache.status.im'...
copying path '/nix/store/mlqss2l2vqqpkr1advdgk1xai8iinm75-androidsdk' from 'https://nix-cache.status.im'...
copying path '/nix/store/n0imzzxf5p1r45j3bx9k5ibspdw2bmqw-android-sdk-build-tools-31.0.0' from 'https://nix-cache.status.im'...
error: unable to download 'https://nix-cache.status.im/nar/ijpabfww24qfcb6fbjkbm184p5pfhm45.nar': HTTP error 200 (curl error: Stream error in the HTTP/2 framing layer)
error: some substitutes for the outputs of derivation '/nix/store/7y0aphnygd0x04y32x3mqsg1aa7hqpzq-androidsdk.drv' failed (usually happens due to networking issues); try '--fallback' to build derivation from source
error: 1 dependencies of derivation '/nix/store/3006mswjsygan8j37cyjqs9q4iim4p84-androidsdk-mod.drv' failed to build
error: build of '/nix/store/3006mswjsygan8j37cyjqs9q4iim4p84-androidsdk-mod.drv', '/nix/store/y5y47bckaqry07zhm06y49xnzvz6kmfb-status-go-0.159.2-fee033f-android.drv' failed
make: *** [Makefile:273: run-android] Error 1
```

jakubgs commented:
As for the Git LFS error:
```
error: builder for '/nix/store/6pr30wwq3nz6vyvpiqc8if652ps5b44l-git-lfs-3.3.0.drv' failed with exit code 1;
panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted [recovered]
```
Why is it even building that, it should come from cache. This error makes no sense.

jakubgs commented:
That Git LFS checks failure is possibly due to use of Rosetta by you, since CI does not use it at all:

* https://github.com/NixOS/nixpkgs/issues/63072

And might be fixed by use of `__darwinAllowLocalNetworking`, but really it shouldn't be happening at all.

jakubgs commented:
The error downloading Android SDK is a well known old issue:
```
error: unable to download 'https://nix-cache.status.im/nar/ijpabfww24qfcb6fbjkbm184p5pfhm45.nar': HTTP error 200 (curl error: Stream error in the HTTP/2 framing layer)
```
You can see it here:

* https://github.com/status-im/infra-ci/issues/17

And it's unrelated to this PR.

siddarthkay commented:
I got past the Git LFS issue with adding an overlay to stop the checks like this in `nix/overlay.nix`

```nix
  git-lfs = super.git-lfs.overrideAttrs (old: {
    __darwinAllowLocalNetworking = true;
    doCheck = false;
  });
```
note :  the   ` __darwinAllowLocalNetworking = true;` bit may not be necessary.

I now see this issue : 

```log
sandbox-exec: pattern serialization length 123946 exceeds maximum (65535)
error: builder for '/nix/store/xrkxs6rgbzwzdfj18jrm9mprf7lyprnv-offline.drv' failed with exit code 65;
       last 1 log lines:
       > sandbox-exec: pattern serialization length 123946 exceeds maximum (65535)
       For full logs, run 'nix log /nix/store/xrkxs6rgbzwzdfj18jrm9mprf7lyprnv-offline.drv'.
error: 1 dependencies of derivation '/nix/store/xwqydmg0553phb8ki8wi9grpw6bp2asj-status-mobile-node-deps-1.23.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/zflgw1gc7q9gpxjyxsf0158v7bqabjvq-status-mobile-node-deps-1.23.0-patched.drv' failed to build
make: *** [run-clojure] Error 1
```

Even after downgrading nix to 2.14.1 the `sandbox-exec` error does not go away.
So this is probably a rosetta thing.
Trying to purge and install nix 2.15.1 in silicon terminal, will report further.

siddarthkay commented:
After purging and re-installing nix on silicon terminal. I again receive `sandbox-exec` issue.
As suggested by @jakubgs it was fixed after setting  `sandbox = false` in `nix/nix.cof`, However a proper solution would be to only disable `sandbox` for `iOS` shells so that the `android` shells can still be sandboxed..

ilmotta commented on nix/scripts/setup.sh:
```diff
@@ -38,6 +40,33 @@ nix_install() {
+
+# Fix for warnings about ignoring binary cache.
+nix_make_user_trusted() {
+    # Singl-user installations do not have this issue.
```
Typo

ilmotta commented on nix/scripts/setup.sh:
```diff
@@ -38,6 +40,33 @@ nix_install() {
+    if grep trusted-users /etc/nix/nix.conf 2>/dev/null; then
+        sed -i "s/trusted-users = \(.*\)$/trusted-users = \1 $USER/" /etc/nix/nix.conf
+    else
+        echo "trusted-users = $USER" | sudo tee -a /etc/nix/nix.conf >/dev/null
```
> Adding a user to trusted-users is essentially equivalent to giving that user root access to the system.
> -- https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-trusted-users

I know you know @jakubgs, but honest question, is there an alternative?

jakubgs replied:
The alternative is to open an issue with Nix and have them fix it. Because honestly I don't know why being a trusted user is a requirement to add a nix cache. It's a new requirement as far as I can tell.

Also, honestly, I don't see a reason why not, as the current user already has root access through sudo on pretty much any developer machine, so I don't see how this is an issue.

ilmotta replied:
> The alternative is to open an issue with Nix and have them fix it. Because honestly I don't know why being a trusted user is a requirement to add a nix cache.

My guess is that it's all related to the fact that trusting a cache of binaries involves some known risks, so maybe they're pushing hard to make it more difficult for casual users (?)

> Also, honestly, I don't see a reason why not, as the current user already has root access through sudo on pretty much any developer machine, so I don't see how this is an issue.

I agree in practice it's fine, but I think ideally developers shouldn't need root access to write mobile apps, but maybe I'm idealizing things too much giving the insane number of dependencies in a RN project.

Anyways, thanks @jakubgs

jakubgs replied:
I agree, you should not need `root` to develop apps, but it is what it is. The one solution could be to allow people to pick which type of installation they want. For MacOS we are forced to use multi-user, but for linux single-user is still an option.

I've changed here the default to be multi-user for linux to be consistent. But I could just leave it as is with single-user install being the default which does not have this issue at all. The other option is to let the user pick at install time.

Also apparently this is intended behavior:

* https://github.com/NixOS/nix/issues/8248

jakubgs replied:
I have moved this change to a separate PR:

* https://github.com/status-im/status-mobile/pull/16357

jakubgs commented:
Apparently the issue with creating profiles and gcroots is resolved in `2.16` according to this comment:
https://discourse.nixos.org/t/upgraded-to-nix-2-14-1-and-profiles-are-not-being-created/29455/3?u=jakubgs

So I will test `2.16.1` to verify this.

jakubgs commented:
I tested `2.16.1` and it doesn't fix the issue:
```
jakubgs@ubuntuvm:~/status-mobile % nix-instantiate --attr shells.default --add-root /nix/var/nix/gcroots/per-user/${USER}/status-mobile/shells.default default.nix
error: creating directory '/nix/var/nix/gcroots/per-user/jakubgs': Permission denied

jakubgs@ubuntuvm:~/status-mobile % nix-instantiate --version
nix-instantiate (Nix) 2.16.1
```
I think I will just roll back the upgrade to `2.13.4` to avoid having to add hack-fixes.

jakubgs commented:
I have reduced this PR to simply downgrade Nix to `2.13.4` to resolve some issues, but not all.

I will make a separate PR for the issue of trusted users.

siddarthkay reviewed (approved):
tested on my mac with `rosetta` terminal, purging and re-installing `nix` works fine on this branch.
